### PR TITLE
Close the socket when authentication expires

### LINKED ISSUE #260: Close the socket when authentication expires
When a websocket requires authentication, and the server uses some sort of token based authentication, the credentials usually have a limited lifetime. The `nes` clients are free to `overrideReconnectionAuth()` and provide new credentials for the case when the websocket drops, however if the websocket does not drop - it still assumes the old credentials on the server side. This means that the websocket is open to send/receive information even though the client may no longer have valid credentials, which could be a security issue in some contexts.

## Proposal

* Registration options should accept an `auth.expiresAt` callback.
* ~~When the authentication endpoint accepts valid credentials, it should `const expiresAt = await auth.expiresAt(request.auth);`.~~ After the client authenticates, the server should `const expiresAt = await auth.expiresAt(credentials, artifacts);`.
* The websocket should be closed on the server side at `expiresAt` time. `undefined` means never expire.
* When the client calls ~~`overrideReconnectionAuth()`~~ `reauthenticate()`, it should also make a request to the authentication endpoint, so that the server can update the `expiresAt` time and extend the websocket lifetime.

### PR DESCRIPTION
**Pull request by dominykas**
Closes #260.

### CODE REVIEW

File: API.md [hunk 1]
```diff
@@ -33,7 +33,8 @@
     - [`await client.subscribe(path, handler)`](#await-clientsubscribepath-handler)
     - [`await client.unsubscribe(path, handler)`](#await-clientunsubscribepath-handler)
     - [`client.subscriptions()`](#clientsubscriptions)
-    - [`client.overrideReconnectionAuth(auth)`](#clientoverriderecinnectionauthauth)
+    - [`client.overrideReconnectionAuth(auth)`](#clientoverridereconnectionauthauth)
+    - [`client.reauthenticate(auth)`](#await-clientreauthenticateauth)
```
Comment (mtharrison) [inline API.md]: Think it should have `await` inside the `[]` too for consistency.

File: API.md [hunk 2]
```diff
@@ -110,6 +111,15 @@ method. The plugin accepts the following optional registration options:
           Defaults to `5000` (5 seconds).
         - `maxConnectionsPerUser` - if specified, limits authenticated users to a maximum number of
           client connections. Requires the `index` option enabled. Defaults to `false`.
+        - `expiresAt` - a function with a signature `async function(auth)`, where `auth` is the
```
Comment (mtharrison) [inline API.md]: Not sure it needs to be an `async` function. In most (all?) cases the expiry info should be available synchronously from the auth?

**Comment by hueniverse** [comment]
This is not the right solution. I don't want nes to have it's own auth setup that's different from hapi. I have another idea.

File: lib/client.js [hunk 3]
```diff
@@ -286,6 +286,18 @@
...
+    Client.prototype.reauthenticate = function (auth) {
+
+        this.overrideReconnectionAuth(auth);
+
+        const request = {
+            type: 'reauth',
+            auth
+        };
+
+        return this._send(request, true).then(() => true);
```
Comment (hueniverse) [inline lib/client.js]: Either we don't care about server response, or we wait for actual response to resolve.

**Comment by dominykas** [comment]
Closing in favor of #262.

**Pull request closed without merging**